### PR TITLE
Added a recipe for base path resolution

### PR DESCRIPTION
**Pull request by weierophinney**
This patch adds a recipe to the cookbook covering how to serve Expressive applications from subdirectories.

Fixes zendframework/zend-expressive-skeleton#51

### CODE REVIEW

**Comment by mtymek** [comment]
Haha, I was about to start writing cookbook now :)

**Comment by weierophinney** [comment]
@mtymek Beat you to it! You can now review this one. :)

File: doc/book/cookbook/using-a-base-path.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        'invokables' => [
+            Blast\BaseUrl\BasePathHelper::class => Blast\BaseUrl\BasePathHelper::class,            
+            /* ... */
+        ],
+        'factories' => [
+            Blast\BaseUrl\UrlHelper::class => Blast\BaseUrl\UrlHelperFactory::class,            
+            /* ... */
+        ],
+        'aliases' => [
+            // alias default UrlHelper with Blast\BaseUrl alternative
```
Comment (mtymek) [inline doc/book/cookbook/using-a-base-path.md]: I just tried it on fresh skeleton installation, and was hit with following error:

`An alias by the name "zendexpressivehelperurlhelper" or "Zend\\Expressive\\Helper\\UrlHelper" already exists'`

Maybe you can indicate that default helper should be removed?

File: doc/book/cookbook/using-a-base-path.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+Finally, if you're using zend-view, you can register a new "basePath" helper in
+your `config/autoload/templates.global.php`:
+
+```
+return [
+    /* ... */
+    'view_helpers' => [
+        'factories' => [
+            'basePath' => Blast\Baseurl\BasePathViewHelperFactory::class,
```
Comment (mtymek) [inline doc/book/cookbook/using-a-base-path.md]: s/Baseurl/BaseUrl

File: doc/book/cookbook/using-a-base-path.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    'dependencies' => [
+        'factories' => [
+            Blast\BaseUrl\BaseUrlMiddleware::class => Blast\BaseUrl\BaseUrlMiddlewareFactory::class,
+            /* ... */
+        ],
+        /* ... */
+    ],
+    'middleware_pipeline' => [
+        'pre_routing' => [
+            [ 'middleware' => [ BaseUrlMiddleware::class ] ],
```
Comment (mtymek) [inline doc/book/cookbook/using-a-base-path.md]: it won't work if copy-pasted directly - maybe class import or FQCN should be used?

File: doc/book/cookbook/using-a-base-path.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        ],
+        /* ... */
+    ],
+];
+```
+
+Usage of the `BasePath` helper is as follows:
+
+```php
+echo $basePathHelper('/icons/favicon.ico');
```
Comment (mtymek) [inline doc/book/cookbook/using-a-base-path.md]: This one won't work directly. Do you think it is enough to leave it like that, assuming reader will understand that he has to pull the helper from the container and use it like this?

**Comment by mtymek** [comment]
Looks good (besides little elements I added) and works well - just tested it.

File: doc/book/cookbook/using-a-base-path.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+```php
+return [
+    'los_basepath' => '<base path here>',
+    /* ... */
+];
+```
+
+The base path should be the portion of the web root leading up to the
+`index.php` of your application. In the above example, this would be
+`/expressive`.
```
Comment (samsonasik) [inline doc/book/cookbook/using-a-base-path.md]: `/expressive/public`

Comment (mtymek) [inline doc/book/cookbook/using-a-base-path.md]: /expressive is correct - see paragraph about .htaccess.

With such approach you don't need to worry about protecting other folders with additional .htaccess files.

Comment (samsonasik) [inline doc/book/cookbook/using-a-base-path.md]: ok :+1:

File: doc/book/cookbook/using-a-base-path.md (hunk 1 above)
Comment (weierophinney) [inline doc/book/cookbook/using-a-base-path.md]: Note has been added now, @mtymek

**Pull request merged**